### PR TITLE
Use neptune implementation of Poseidon

### PR DESCRIPTION
**Pull request by jonas-lj**
Use a [fork](https://github.com/jonas-lj/neptune) of lurk lab's [poseidon implementation](https://github.com/lurk-lab/neptune) instead of poseidon-rs. This gives a 2x improvement for 4 inputs and a 5x improvement for >= 8 inputs.

It's necessary to upgrade the rust toolchain version, which introduced a few new clippy warnings which is also handled.

This gives a 25% improvement in verification of zkLogin signatures (2 ms -> 1.5 ms).

Note: The neptune fork is being merged into main (https://github.com/lurk-lab/neptune/pull/236), so let's merge this PR after that.

### CODE REVIEW

File: fastcrypto-vdf/src/class_group/gmp/compressed.rs [hunk 1]
```diff
@@ -36,12 +36,12 @@ impl QuadraticForm {
     /// length in bits.
     pub fn serialized_length(discriminant_in_bits: usize) -> usize {
         // The number of 32 bit words needed to represent the discriminant rounded up,
-        ((discriminant_in_bits + 31) / 32
```
Comment (jonas-lj) [inline fastcrypto-vdf/src/class_group/gmp/compressed.rs:39]: After updating the rust toolchain, clippy gives new warnings.

File: fastcrypto-tbls/src/tests/nodes_tests.rs [hunk 2]
```diff
@@ -16,7 +16,7 @@ fn test_reduce() {
             .map(|i| Node {
                 id: i,
                 pk: pk.clone(),
-                weight: 5 + i as u16,
```
Comment (jonas-lj) [inline fastcrypto-tbls/src/tests/nodes_tests.rs:19]: After updating the rust toolchain, clippy gives new warnings.

**Comment by huitseeker** [comment]
differential CPU bench on M1 at 9ff2263:
```
group                                                                    jonas-lj                               main
-----                                                                    --------                               ----
arity-11/bls                                                             1.00     48.8±0.51µs            1.01     49.1±0.85µs
arity-11/pallas                                                          1.00     65.9±0.83µs            1.01     66.5±0.75µs
arity-11/vesta                                                           1.00     65.5±0.58µs            1.02     66.9±0.97µs
arity-2/bls                                                              1.02     11.1±0.20µs            1.00     11.0±0.13µs
arity-2/pallas                                                           1.00     14.4±0.16µs            1.01     14.5±0.23µs
arity-2/vesta                                                            1.02     14.6±0.26µs            1.00     14.3±0.19µs
arity-4/bls                                                              1.00     17.7±0.36µs            1.01     17.9±0.25µs
arity-4/pallas                                                           1.00     23.9±0.25µs            1.00     23.8±0.28µs
arity-4/vesta                                                            1.00     23.5±0.26µs            1.00     23.5±0.26µs
arity-8/bls                                                              1.01     34.1±0.39µs            1.00     33.8±0.34µs
arity-8/pallas                                                           1.00     45.8±0.46µs            1.02     46.6±0.73µs
arity-8/vesta                                                            1.00     45.7±0.38µs            1.02     46.7±0.65µs
hash-bls-128/Poseidon hash optimized (strengthened)/Generated scalars    1.01     21.8±0.23µs            1.00     21.5±0.23µs
hash-bls-128/Poseidon hash optimized/Generated scalars                   1.01     18.8±0.24µs            1.00     18.6±0.25µs
hash-bls-128/Poseidon hash/Generated scalars                             1.01     32.9±0.36µs            1.00     32.7±0.47µs
hash-bls-128/Sha2 256/Generated scalars                                  1.00  1547.6±25.37ns            1.00  1554.8±29.92ns
hash-bls-128/Sha2 512/Generated scalars                                  1.00  1457.5±14.33ns            1.01  1467.8±18.13ns
hash-bls-256/Poseidon hash optimized (strengthened)/Generated scalars    1.00     41.5±0.63µs            1.00     41.4±0.56µs
hash-bls-256/Poseidon hash optimized/Generated scalars                   1.02     36.7±0.64µs            1.00     36.1±0.51µs
hash-bls-256/Poseidon hash/Generated scalars                             1.02     93.6±1.23µs            1.00     91.5±1.06µs
hash-bls-256/Sha2 256/Generated scalars                                  1.01      2.9±0.05µs            1.00      2.9±0.04µs
hash-bls-256/Sha2 512/Generated scalars                                  1.00      2.7±0.03µs            1.00      2.7±0.03µs
hash-bls-352/Poseidon hash optimized (strengthened)/Generated scalars    1.00     57.7±0.63µs            1.02     59.0±1.05µs
hash-bls-352/Poseidon hash optimized/Generated scalars                   1.02     52.4±0.98µs            1.00     51.4±0.60µs
hash-bls-352/Poseidon hash/Generated scalars                             1.00    154.8±1.96µs            1.00    155.1±1.72µs
hash-bls-352/Sha2 256/Generated scalars                                  1.00      3.8±0.04µs            1.00      3.8±0.06µs
hash-bls-352/Sha2 512/Generated scalars                                  1.00      3.4±0.04µs            1.00      3.4±0.04µs
hash-bls-64/Poseidon hash optimized (strengthened)/Generated scalars     1.02     13.6±0.24µs            1.00     13.4±0.15µs
hash-bls-64/Poseidon hash optimized/Generated scalars                    1.02     11.5±0.12µs            1.00     11.3±0.12µs
hash-bls-64/Poseidon hash/Generated scalars                              1.02     15.5±0.17µs            1.00     15.2±0.19µs
hash-bls-64/Sha2 256/Generated scalars                                   1.00    868.2±9.89ns            1.01   878.1±10.75ns
hash-bls-64/Sha2 512/Generated scalars                                   1.00    746.9±9.68ns            1.00   746.6±11.06ns
```

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs [hunk 3]
```diff
@@ -1,65 +1,82 @@
...
+        for input in $inputs.iter() {
+            poseidon.input(bn254_to_fr(*input)).unwrap();
+        }
+        poseidon.hash_in_mode(OptimizedStatic);
 
-impl Debug for PoseidonWrapper {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("PoseidonWrapper").finish()
-    }
+        // Neptune returns the state element with index 1 but we want the first element.
```
Comment (kchalkias) [inline fastcrypto-zkp/src/bn254/poseidon/mod.rs:35]: Can you elaborate a bit more? What is the rationale between the two choices?

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs [hunk 4]
```diff
@@ -1,65 +1,82 @@
...
+        poseidon.elements[0]
+    }};
 }
 
-impl Default for PoseidonWrapper {
-    fn default() -> Self {
-        Self::new()
+/// Poseidon hash function over BN254.
+pub fn hash(inputs: Vec<Fr>) -> Result<Fr, FastCryptoError> {
```
Comment (kchalkias) [inline fastcrypto-zkp/src/bn254/poseidon/mod.rs:41]: So it's a restricted size input hash function, right? and we guarantee we have 16 constants to accommodate it?
update: ah is it because we create chunks of 16 elements when we have more? and then hash them all together?

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs [hunk 5]
```diff
@@ -1,65 +1,82 @@
...
-    }
-
-    /// Calculate the hash of the given inputs.
-    pub fn hash(&self, inputs: Vec<Fr>) -> Result<Fr, FastCryptoError> {
-        self.instance
-            .hash(inputs)
-            .map_err(|_| FastCryptoError::InvalidInput)
-    }
+    // Instances of Poseidon and PoseidonConstants from neptune have different types depending on
+    // the number of inputs, so unfortunately we need to use a macro here.
```
Comment (kchalkias) [inline fastcrypto-zkp/src/bn254/poseidon/mod.rs:47]: +1

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs [hunk 6]
```diff
@@ -1,65 +1,82 @@
...
 /// Calculate the poseidon hash of the field element inputs. If the input length is <= 16, calculate
 /// H(inputs), if it is <= 32, calculate H(H(inputs[0..16]), H(inputs[16..])), otherwise return an
 /// error.
 pub fn to_poseidon_hash(inputs: Vec<Fr>) -> Result<Fr, FastCryptoError> {
-    static POSEIDON: OnceCell<PoseidonWrapper> = OnceCell::new();
-    let poseidon_ref = POSEIDON.get_or_init(PoseidonWrapper::new);
     if inputs.len() <= 16 {
-        poseidon_ref.hash(inputs)
+        hash(inputs)
     } else if inputs.len() <= 32 {
```
Comment (kchalkias) [inline fastcrypto-zkp/src/bn254/poseidon/mod.rs:76]: So 32 is the max? and if someone has a longer input, they should split beforehands and recursively invoke that function? Do we want to offer that recursive (for loop) function, to accommodate longer inputs as well (where internally it does the splits and merge for any size of inputs?

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs [hunk 7]
```diff
@@ -223,21 +254,20 @@ mod test {
...
-                payload_len,
-                eph_public_key_0,
-                eph_public_key_1,
-                max_epoch,
-                num_sha2_blocks,
-                key_claim_name_f,
-                addr_seed,
-            ])
-            .unwrap();
+        let hash = hash(vec![
```
Comment (kchalkias) [inline fastcrypto-zkp/src/bn254/poseidon/mod.rs:257]: is this a zkLogin input test? if yes why it's not in the zkLogin test file? is it because it happens to have the number of inputs we expect for this test?

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs [hunk 8]
```diff
@@ -271,4 +301,25 @@ mod test {
         let inputs = vec![vec![255; 31]];
         assert!(hash_to_bytes(&inputs).is_ok());
     }
+
+    #[cfg(test)]
```
Comment (kchalkias) [inline fastcrypto-zkp/src/bn254/poseidon/mod.rs:305]: are there any official test vectors or can we fuzzy test against another Poseidon lib (ie the previous one?)?

File: fastcrypto-zkp/Cargo.toml [hunk 9]
```diff
@@ -39,11 +39,14 @@ schemars ="0.8.10"
 serde = { version = "1.0.152", features = ["derive"] }
 serde_json = "1.0.93"
 once_cell = "1.16"
-poseidon-ark = { git = "https://github.com/arnaucube/poseidon-ark.git", rev = "ff7f5e05d55667b4ffba129b837da780c4c5c849" }
 bcs = "0.1.4"
 im = "15"
 reqwest = { version = "0.11.20", default_features = false, features = ["blocking", "json", "rustls-tls"] }
 rustls-webpki = "0.101.4"
+neptune = { git = "https://github.com/lurk-lab/neptune", rev = "d8b4eeadd8acc9d9e8d9d510605c954f1410aa60" }
```
Comment (kchalkias) [inline fastcrypto-zkp/Cargo.toml:46]: it's now merged in main branch?

Comment (jonas-lj) [reply fastcrypto-zkp/Cargo.toml:46]: Yes. I just pinned it to a commit until it's released.

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs (hunk 6 above)
Comment (jonas-lj) [reply fastcrypto-zkp/src/bn254/poseidon/mod.rs:76]: We could definitely do that. This design is aligned with the TS code due to zklogin. I think we should keep it as is for now and then document it more clearly.

Comment (kchalkias) [reply fastcrypto-zkp/src/bn254/poseidon/mod.rs:76]: @jonas-lj can we at least document that this is the case and we plan to update in the future? If used generically (ie in smart contracts) they should know the limits.

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs (hunk 3 above)
Comment (jonas-lj) [reply fastcrypto-zkp/src/bn254/poseidon/mod.rs:35]: I've added some links to where the output is extracted from the state in neptune, poseidon-rs and circomlib's implementation. We're aligned with the latter two.

The paper doesn't state anything about which element to use. They define the hash function on n elements as a permutation, so it shouldn't matter which element you use.

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs (hunk 4 above)
Comment (jonas-lj) [reply fastcrypto-zkp/src/bn254/poseidon/mod.rs:41]: Yes. Up to 32 elements. This design is a bit arbitrary though, but is aligned with the circuit.

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs (hunk 8 above)
Comment (jonas-lj) [reply fastcrypto-zkp/src/bn254/poseidon/mod.rs:305]: There are [test vectors](https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/test_vectors.txt?ref_type=heads) which we're aligned with, but I don't know any test vectors for edge cases.

I've added a proptest to verify alignment with poseidon-rs for 1-16 inputs, but it's not checking edge cases. I'll look into what options we have.

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs (hunk 7 above)
Comment (jonas-lj) [reply fastcrypto-zkp/src/bn254/poseidon/mod.rs]: Good point. I've moved the test.

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs (hunk 6 above)
Comment (jonas-lj) [reply fastcrypto-zkp/src/bn254/poseidon/mod.rs:83]: Sure. Now that I think more about it, we could also extend it to more inputs now by using a Merkle tree with degree 16. Then it'll me aligned with the circuit for <= 32 inputs but still allow for more.

**Review by joyqvq — APPROVED** [review]
lgtm

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs [hunk 10]
```diff
@@ -1,65 +1,89 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
+use crate::bn254::poseidon::constants::{
```
Comment (kchalkias) [inline fastcrypto-zkp/src/bn254/poseidon/mod.rs:4]: Does fmt complain? if we order them properly? alternatively we could use U01, U02 etc. In any case, minor.

**Review by kchalkias — APPROVED** [review]

File: fastcrypto-zkp/src/bn254/poseidon/mod.rs (hunk 10 above)
Comment (jonas-lj) [reply fastcrypto-zkp/src/bn254/poseidon/mod.rs]: I did the quick-fix of replacing the imports with a `*`.

**Comment by huitseeker** [comment]
FYI, you should now be able to rely on https://crates.io/crates/neptune/13.0.0 which includes your changes.

**Comment by jonas-lj** [comment]
> @jonas-lj FYI, you should now be able to rely on https://crates.io/crates/neptune/13.0.0 which includes your changes.

Nice - thanks!
